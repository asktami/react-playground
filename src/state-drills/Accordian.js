import React from 'react';
import './Accordian.css';

// display NO sections by default
// only show content when the sections array has length
// show the last section clicked
// only show 1 section at a time

export class Accordian extends React.Component {
	// define default sections empty array
	// any 'sections' props value is passed from App.js will be used instead
	static defaultProps = {
		sections: []
	};

	// default array index/section/button clicked/currentSectionIndex = 0
	state = {
		currentSectionIndex: null
	};

	toggleActive = index => e => {
		// remove active class from anything that has active class
		// then add active class to the selected button
		if (document.querySelector('.active') !== null) {
			document.querySelector('.active').classList.remove('active');
		}
		index.currentTarget.classList.add('active');
	};

	// property initializer syntax
	// it binds the event handler to the class AND lets you pass an argument (index) to the function/method on event (e) or ()
	handleButtonClick = index => event => {
		this.setState({ currentSectionIndex: index });

		// QUESTION - how to get toggleActive to work???
		// this.toggleActive(index);

		// remove active class from anything that has active class
		// then add active class to the selected button
		if (document.querySelector('.active') !== null) {
			document.querySelector('.active').classList.remove('active');
		}
		event.currentTarget.classList.add('active');
	};

	renderButtons() {
		// VERTICAL LIST OF BUTTONS
		const { sections } = this.props; // because only 1 props property
		const { currentSectionIndex } = this.state; // because only 1 state property

		const buttonList = sections.map((section, index) => (
			<li key={index}>
				<button className="accordion" onClick={this.handleButtonClick(index)}>
					{section.title}
				</button>

				{/*
				// render content only for clicked button
				// only clicked button section should open, close all others
				// not not this.props.section.length === if this.props.section.length > 0 */}
				{!!sections.length &&
					currentSectionIndex === index &&
					this.renderContent(sections, currentSectionIndex)}
			</li>
		));

		return <ul className="accordianList">{buttonList}</ul>;

		/*
		// HORIZONTAL LIST OF BUTTONS
		return sections.map((section, index) => (
			<button key={index} onClick={this.handleButtonClick(index)}>
				{tab.title}
			</button>
		));
		*/
	}

	renderContent(sections, currentSectionIndex) {
		// show section content for button clicked
		const currentSection = sections[currentSectionIndex];
		return (
			<div className="panel">
				<p>{currentSection.content}</p>
			</div>
		);
	}

	render() {
		return <div>{this.renderButtons()}</div>;
	}
}
