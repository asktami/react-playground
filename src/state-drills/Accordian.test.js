// using only Enzyme for testing, not Jest (react-test-renderer)

import React from 'react';

import { shallow } from 'enzyme'; // add enzyme for event listener tests
import toJson from 'enzyme-to-json'; // to create a JSON version of the Jest wrapper

import { Accordian } from './Accordian';

// Accordian component sections prop
const sectionsProp = [
	{
		title: 'Section 1',
		content:
			'Section 1: Lorem ipsum dolor sit amet consectetur adipisicing elit.'
	},
	{
		title: 'Section 2',
		content:
			'Section 2: Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!'
	},
	{
		title: 'Section 3',
		content:
			'Section 3: Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?'
	}
];

describe(`Accordian Component`, () => {
	// smoke test
	// renders empty given no Accordion without errors
	it('renders an empty li when the sections prop is not supplied', () => {
		const wrapper = shallow(<Accordian />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('renders no sections by default', () => {
		const wrapper = shallow(<Accordian sections={sectionsProp} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	// test for clicking on one of the buttons
	// .at(1) = button at array index position 1 (i.e. 2nd button)
	it('opens one section at a time, the clicked section, open section 2', () => {
		const wrapper = shallow(<Accordian sections={sectionsProp} />);
		wrapper
			.find('button')
			.at(2)
			.simulate('click');
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	// test for clicking on one of the buttons
	// only open the last section clicked
	it('only open the last section clicked, open section 3', () => {
		const wrapper = shallow(<Accordian sections={sectionsProp} />);
		wrapper
			.find('button')
			.at(1)
			.simulate('click');
		wrapper
			.find('button')
			.at(2)
			.simulate('click');
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});
