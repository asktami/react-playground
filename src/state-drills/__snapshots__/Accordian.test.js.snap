// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordian Component only open the last section clicked, open section 3 1`] = `
<div>
  <ul
    className="accordianList"
  >
    <li
      key="0"
    >
      <button
        className="accordion"
        onClick={[Function]}
      >
        Section 1
      </button>
    </li>
    <li
      key="1"
    >
      <button
        className="accordion"
        onClick={[Function]}
      >
        Section 2
      </button>
    </li>
    <li
      key="2"
    >
      <button
        className="accordion"
        onClick={[Function]}
      >
        Section 3
      </button>
      <div
        className="panel"
      >
        <p>
          Section 3: Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?
        </p>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Accordian Component opens one section at a time, the clicked section, open section 2 1`] = `
<div>
  <ul
    className="accordianList"
  >
    <li
      key="0"
    >
      <button
        className="accordion"
        onClick={[Function]}
      >
        Section 1
      </button>
    </li>
    <li
      key="1"
    >
      <button
        className="accordion"
        onClick={[Function]}
      >
        Section 2
      </button>
    </li>
    <li
      key="2"
    >
      <button
        className="accordion"
        onClick={[Function]}
      >
        Section 3
      </button>
      <div
        className="panel"
      >
        <p>
          Section 3: Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?
        </p>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Accordian Component renders an empty li when the sections prop is not supplied 1`] = `
<div>
  <ul
    className="accordianList"
  />
</div>
`;

exports[`Accordian Component renders no sections by default 1`] = `
<div>
  <ul
    className="accordianList"
  >
    <li
      key="0"
    >
      <button
        className="accordion"
        onClick={[Function]}
      >
        Section 1
      </button>
    </li>
    <li
      key="1"
    >
      <button
        className="accordion"
        onClick={[Function]}
      >
        Section 2
      </button>
    </li>
    <li
      key="2"
    >
      <button
        className="accordion"
        onClick={[Function]}
      >
        Section 3
      </button>
    </li>
  </ul>
</div>
`;
